### PR TITLE
wip: implement search cache

### PR DESCRIPTION
**Pull request by yogimathius**
### :tophat: Instructions

### CODE REVIEW

File: src/frontend_cache/cache.c [hunk 1]
```diff
@@ -1,6 +1,35 @@
 
 #include "../../inc/main.h"
 
+t_cache* create_frontend_cache(t_pagemap* pagemap) {
```
Comment (yogimathius) [inline src/frontend_cache/cache.c:4]: @rswhipple I feel like having this function at the top helps orient other developers to the main purpose of this component of our work, which builds a solid foundation for the search functions below. Hope you don't mind me moving your `search_fast_cache` for that reason!

File: src/frontend_cache/cache.c [hunk 2]
```diff
@@ -25,9 +54,46 @@ void* search_fast_cache(size_t size) {
     return NULL;
 }
 
+void* search_unsorted_cache(size_t size) {
+    t_chunk* unsorted_cache = g_pagemap->frontend_cache->unsorted_cache;
```
Comment (yogimathius) [inline src/frontend_cache/cache.c:58]: definitely more work to do in this function, as you were saying it should sort chunks into the cache_table!

I'll spend more time this weekend to look into that work

File: src/frontend_cache/cache.c [hunk 3]
```diff
@@ -25,9 +54,46 @@ void* search_fast_cache(size_t size) {
...
+
+    return NULL;
+}
+
+
+void* search_sorted_cache(size_t size) {
+    char key[32];
+    snprintf(key, sizeof(key), "%zu", size);
+
+    cache_table* cache_table = g_pagemap->frontend_cache->cache_table;
```
Comment (yogimathius) [inline src/frontend_cache/cache.c:75]: still need to determine how the pagetype fits with the cache_table, thoughts? @rswhipple

**Comment by yogimathius** [comment]
this is marked as a WIP but please feel free to merge it into your branch when you'd like to, and I think we can merge yours into main and branch out again so we can keep working from a clean state of main?

**Comment by yogimathius** [comment]
honestly not even sure if [this is how we're supposed to be using top_chunk ](https://github.com/rswhipple/jsmalloc/pull/45/commits/fb1c888e7e81fe59ca1433385f488aee82183231)😬 just getting some sort of implementation to roll with and get a better understanding of the chunks

File: src/frontend_cache/cache.c (hunk 1 above)
Comment (rswhipple) [reply src/frontend_cache/cache.c:4]: no problem, I'm often on the fence as to which function should go first

File: src/frontend_cache/cache.c (hunk 3 above)
Comment (rswhipple) [reply src/frontend_cache/cache.c:75]: I think it's mostly aspirational for this implementation. With a next version we could implement tailored algorithms to handle small vs large chunks. But, for this version, we could at minimum have them creating new chunks from the correct heap (small vs large)?

**Review by rswhipple — APPROVED** [review]
awesome, let's keep merging and iterating, getting close! I'll start a PR for my branch, too, so we can get main up to date

**Pull request merged**